src/python/pants/backend/python/goals/setup_py.py

```diff
@@ -535,33 +536,45 @@ class GeneratedSetupPy:
 
 @rule
 async def generate_chroot(request: SetupPyChrootRequest) -> SetupPyChroot:
+    generate_setup = request.exported_target.target.get(GenerateSetupField).value
+
+    # TODO: In 2.9.0.dev0 delete from here...
+    resolved_setup_kwargs = await Get(SetupKwargs, ExportedTarget, request.exported_target)
+    setup_script = resolved_setup_kwargs.kwargs.get("setup_script")
+    if setup_script:
+        generate_setup = False
+        warn_or_error(
+            "2.9.0.dev0",
+            "explicitly specifying a setup_script",
+            f"This kwarg is ignored. Instead set generate_setup=False on the python_distribution "
+            f"target at {request.exported_target.target.address.spec}, and setup will run on "
+            f"whatever relevant files it finds in the dependency closure (e.g., setup.py, "
+            f"setup.cfg, pyproject.toml",
+        )
+    # ... to here.
+
     sources = await Get(SetupPySources, SetupPyChrootRequest, request)
-    snapshot = await Get(Snapshot, Digest, sources.digest)
-    stripped = await Get(
-        StrippedSourceFileNames,
-        SourcesPaths(
-            files=(os.path.join(request.exported_target.target.address.spec_path, "setup.py"),),
-            dirs=(),
-        ),
-    )
-    existing_setup_py = stripped[0]
 
-    # TODO: Add a generate_setup field to the python_distribution target type, so that
-    #  we're not guessing user intent (and not relying on the build being specifically
-    #  setup.py-based, it could be setup.cfg).
-    if existing_setup_py in snapshot.files:
-        working_directory = os.path.dirname(existing_setup_py)
-        chroot_digest = sources.digest
-    else:
-        logger.info("No setup.py found, generating one...")
+    if generate_setup:
         generated_setup_py = await Get(
             GeneratedSetupPy, GenerateSetupPyRequest(request.exported_target, sources)
         )
         # We currently generate a setup.py that expects to be in the source root.
         # TODO: It might make sense to generate one in the target's directory, for
-        #  consistency with existing setup.py.
+        #  consistency with the existing setup.py case.
         working_directory = ""
         chroot_digest = await Get(Digest, MergeDigests((sources.digest, generated_setup_py.digest)))
+    else:
+        # To get the stripped target directory we need a dummy path under it.
+        stripped = await Get(
+            StrippedSourceFileNames,
+            SourcesPaths(
+                files=(os.path.join(request.exported_target.target.address.spec_path, "dummy.py"),),
```

Comment:
You could get unlucky here IIUC. It might be good to fully be sure - maybe use a uuid - or else tag this for a follow up rework to not need the hack.

Reply:
How so? We don't write this file, it's just a string we use for getting the source-root stripped dir, because our source root stripping mechanisms assume that the last path component is a filename, and so start the search from its parent.

Reply:
Added a comment and TODO to clarify.